include/ringmesh/mesh/surface_mesh.h

```diff
@@ -0,0 +1,402 @@
+        bool is_polygon_on_border( index_t polygon_index ) const;
+
+        /*!
+         * @brief Gets the length of the edge starting at a given vertex
+         * @param[in] polygon_local_edge index of the polygon and its
+         * local edge starting vertex index
+         */
+        double polygon_edge_length(
+            const PolygonLocalEdge& polygon_local_edge ) const;
+
+        /*!
+         * @brief Gets the barycenter of the edge starting at a given vertex
+         * @param[in] polygon_local_edge index of the polygon and
+         * the local edge starting vertex index
+         */
+        vecn< DIMENSION > polygon_edge_barycenter(
+            const PolygonLocalEdge& polygon_local_edge ) const;
+
+        /*!
+         * @brief Gets the vertex index on the polygon edge
+         * @param[in] polygon_local_edge index of the polygon and
+         * the local index of the edge in the polygon
+         * @param[in] vertex_id index of the local vertex in the edge \param
+         * edge_id (0 or 1)
+         * @return the vertex index
+         */
+        index_t polygon_edge_vertex( const PolygonLocalEdge& polygon_local_edge,
+            index_t vertex_id ) const;
+
+        /*!
+         * Computes the Mesh polygon barycenter
+         * @param[in] polygon_id the polygon index
+         * @return the polygon center
+         */
+        vecn< DIMENSION > polygon_barycenter( index_t polygon_id ) const;
+
+        /*!
+         * Computes the Mesh polygon area
+         * @param[in] polygon_id the polygon index
+         * @return the polygon area
+         */
+        virtual double polygon_area( index_t polygon_id ) const = 0;
+
+        /*!
+         * @brief return the NNSearch at polygons
+         */
+        const NNSearch< DIMENSION >& polygon_nn_search() const
+        {
+            if( !nn_search_ )
+            {
+                std::vector< vecn< DIMENSION > > polygon_centers(
+                    nb_polygons() );
+                for( auto p : range( nb_polygons() ) )
+                {
+                    polygon_centers[p] = polygon_barycenter( p );
+                }
+                nn_search_.reset(
+                    new NNSearch< DIMENSION >( polygon_centers, true ) );
+            }
+            return *nn_search_.get();
+        }
+        /*!
+         * @brief Creates an AABB tree for a Mesh polygons
+         */
+        const SurfaceAABBTree< DIMENSION >& polygon_aabb() const
+        {
+            if( !polygon_aabb_ )
+            {
+                polygon_aabb_.reset(
+                    new SurfaceAABBTree< DIMENSION >( *this ) );
+            }
+            return *polygon_aabb_;
+        }
+
+        bool is_mesh_valid() const override;
+
+        std::tuple< index_t, std::vector< index_t > >
+            connected_components() const final;
+
+    protected:
+        SurfaceMeshBase() = default;
+
+    protected:
+        mutable std::unique_ptr< NNSearch< DIMENSION > > nn_search_{};

```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Make this member variable "private". [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=cpp%3AS3656)